File: backend/src/persistence/catalog.rs
use std::fs;
use std::path::PathBuf;

use chrono::Utc;
use serde::{Deserialize, Serialize};

use crate::api::{MeasurementScale, StudyRecord};
use crate::error::{BackendError, BackendResult};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RecentStudyEntry {
    pub input_path: PathBuf,
    pub input_name: String,
    pub measurement_scale: Option<PersistedMeasurementScale>,
    pub last_opened_at: String,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StudyCatalog {
    pub recent_studies: Vec<RecentStudyEntry>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PersistedMeasurementScale {
    pub row_spacing_mm: f64,
    pub column_spacing_mm: f64,
    pub source: String,
}

impl From<MeasurementScale> for PersistedMeasurementScale {
    fn from(scale: MeasurementScale) -> Self {
        Self {
            row_spacing_mm: scale.row_spacing_mm,
            column_spacing_mm: scale.column_spacing_mm,
            source: scale.source.to_string(),
        }
    }
}

#[derive(Debug, Clone)]
pub struct StudyCatalogStore {
    path: PathBuf,
}

impl StudyCatalogStore {
    pub fn new(path: PathBuf) -> Self {
        Self { path }
    }

    pub fn load_or_default(&self) -> StudyCatalog {
        self.load().unwrap_or_default()
    }

    pub fn load(&self) -> BackendResult<StudyCatalog> {
        let Ok(contents) = fs::read_to_string(&self.path) else {
            return Ok(StudyCatalog::default());
        };

        serde_json::from_str(&contents).map_err(|error| {
            let _ = fs::rename(&self.path, self.path.with_extension("corrupt.json"));
            BackendError::cache_corrupted(format!(
                "study catalog at {} is invalid JSON: {error}",
                self.path.display()
            ))
        })
    }

    pub fn record_opened_study(&self, study: &StudyRecord) -> BackendResult<()> {
        let mut catalog = self.load_or_default();
        catalog
            .recent_studies
            .retain(|entry| entry.input_path != study.input_path);
        catalog.recent_studies.insert(
            0,
            RecentStudyEntry {
                input_path: study.input_path.clone(),
                input_name: study.input_name.clone(),
                measurement_scale: study.measurement_scale.map(Into::into),
                last_opened_at: Utc::now().to_rfc3339(),
            },
        );
        catalog.recent_studies.truncate(10);
        self.save(&catalog)
    }

    fn save(&self, catalog: &StudyCatalog) -> BackendResult<()> {
        if let Some(parent) = self.path.parent() {
            fs::create_dir_all(parent).map_err(|error| {
                BackendError::internal(format!(
                    "failed to create catalog directory {}: {error}",
                    parent.display()
                ))
            })?;
        }

        let payload = serde_json::to_string_pretty(catalog)
            .map_err(|error| BackendError::internal(format!("serialize study catalog: {error}")))?;
        fs::write(&self.path, payload).map_err(|error| {
            BackendError::internal(format!(
                "failed to write study catalog {}: {error}",
                self.path.display()
            ))
        })
    }
}

#[cfg(test)]
mod tests {
    use std::fs;
    use tempfile::tempdir;

    use crate::api::StudyRecord;

    use super::StudyCatalogStore;

    #[test]
    fn record_opened_study_keeps_most_recent_entry_first() {
        let temp = tempdir().expect("temp dir");
        let store = StudyCatalogStore::new(temp.path().join("catalog.json"));

        store
            .record_opened_study(&StudyRecord {
                study_id: String::from("study-1"),
                input_path: "/tmp/one.dcm".into(),
                input_name: String::from("one.dcm"),
                measurement_scale: None,
            })
            .expect("record first study");
        store
            .record_opened_study(&StudyRecord {
                study_id: String::from("study-2"),
                input_path: "/tmp/two.dcm".into(),
                input_name: String::from("two.dcm"),
                measurement_scale: None,
            })
            .expect("record second study");

        let catalog = store.load().expect("load catalog");
        assert_eq!(catalog.recent_studies[0].input_name, "two.dcm");
        assert_eq!(catalog.recent_studies[1].input_name, "one.dcm");
    }

    #[test]
    fn invalid_catalog_is_treated_as_corrupted_cache() {
        let temp = tempdir().expect("temp dir");
        let catalog_path = temp.path().join("catalog.json");
        fs::write(&catalog_path, "{ not json").expect("write invalid catalog");
        let store = StudyCatalogStore::new(catalog_path.clone());

        let error = store.load().expect_err("invalid catalog should error");

        assert_eq!(error.code, crate::error::BackendErrorCode::CacheCorrupted);
        assert!(catalog_path.with_extension("corrupt.json").exists());
    }
}

File: backend/internal/persistence/catalog.go
package persistence

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"

	"xrayview/backend/internal/contracts"
)

const recentStudyLimit = 10

type RecentStudyEntry struct {
	InputPath        string                     `json:"inputPath"`
	InputName        string                     `json:"inputName"`
	MeasurementScale *PersistedMeasurementScale `json:"measurementScale"`
	LastOpenedAt     string                     `json:"lastOpenedAt"`
}

type StudyCatalog struct {
	RecentStudies []RecentStudyEntry `json:"recentStudies"`
}

type PersistedMeasurementScale struct {
	RowSpacingMM    float64 `json:"rowSpacingMm"`
	ColumnSpacingMM float64 `json:"columnSpacingMm"`
	Source          string  `json:"source"`
}

type Catalog struct {
	rootDir string
	path    string
	now     func() time.Time

	mu     sync.Mutex
	loaded bool
	cache  StudyCatalog
}

func New(rootDir string) *Catalog {
	return NewAtPath(filepath.Join(filepath.Clean(rootDir), "catalog.json"))
}

func NewAtPath(path string) *Catalog {
	cleanPath := filepath.Clean(path)

	return &Catalog{
		rootDir: filepath.Dir(cleanPath),
		path:    cleanPath,
		now: func() time.Time {
			return time.Now().UTC()
		},
	}
}

func (catalog *Catalog) RootDir() string {
	return catalog.rootDir
}

func (catalog *Catalog) Path() string {
	return catalog.path
}

func (catalog *Catalog) Ensure() error {
	return os.MkdirAll(catalog.rootDir, 0o755)
}

// Load reads the persisted study catalog. A missing file is the expected
// first-run state — we swallow the error and return an empty catalog so
// nothing upstream has to special-case it. Corrupt JSON is different:
// we rename the bad file to *.corrupt first, then surface the error.
// That keeps the next save() from silently overwriting the evidence.
func (catalog *Catalog) Load() (StudyCatalog, error) {
	catalog.mu.Lock()
	defer catalog.mu.Unlock()

	value, err := catalog.loadFromDisk()
	if err != nil {
		catalog.loaded = false
		catalog.cache = emptyStudyCatalog()
		return value, err
	}

	catalog.cache = cloneStudyCatalog(value)
	catalog.loaded = true

	return cloneStudyCatalog(value), nil
}

func (catalog *Catalog) loadFromDisk() (StudyCatalog, error) {
	contents, err := os.ReadFile(catalog.path)
	if err != nil {
		if os.IsNotExist(err) {
			return emptyStudyCatalog(), nil
		}

		return emptyStudyCatalog(), contracts.Internal(
			fmt.Sprintf("failed to read study catalog %s: %v", catalog.path, err),
		)
	}

	var value StudyCatalog
	if err := json.Unmarshal(contents, &value); err != nil {
		_ = os.Rename(catalog.path, catalog.corruptPath())
		return emptyStudyCatalog(), contracts.CacheCorrupted(
			fmt.Sprintf("study catalog at %s is invalid JSON: %v", catalog.path, err),
		)
	}
	if value.RecentStudies == nil {
		value.RecentStudies = []RecentStudyEntry{}
	}

	return value, nil
}

func (catalog *Catalog) RecordOpenedStudy(study contracts.StudyRecord) error {
	catalog.mu.Lock()
	defer catalog.mu.Unlock()

	value, err := catalog.loadOrDefaultLocked()
	if err != nil {
		return err
	}

	filtered := make([]RecentStudyEntry, 0, len(value.RecentStudies))
	for _, entry := range value.RecentStudies {
		if entry.InputPath != study.InputPath {
			filtered = append(filtered, entry)
		}
	}

	entry := RecentStudyEntry{
		InputPath:        study.InputPath,
		InputName:        study.InputName,
		MeasurementScale: persistedMeasurementScale(study.MeasurementScale),
		LastOpenedAt:     catalog.now().Format(time.RFC3339),
	}
	value.RecentStudies = append([]RecentStudyEntry{entry}, filtered...)
	if len(value.RecentStudies) > recentStudyLimit {
		value.RecentStudies = value.RecentStudies[:recentStudyLimit]
	}

	return catalog.save(value)
}

func (catalog *Catalog) loadOrDefaultLocked() (StudyCatalog, error) {
	if catalog.loaded {
		return cloneStudyCatalog(catalog.cache), nil
	}

	value, err := catalog.loadFromDisk()
	if err != nil {
		if backendErr, ok := err.(contracts.BackendError); ok &&
			backendErr.Code == contracts.BackendErrorCodeCacheCorrupted {
			return emptyStudyCatalog(), nil
		}

		return emptyStudyCatalog(), err
	}

	catalog.cache = cloneStudyCatalog(value)
	catalog.loaded = true

	return cloneStudyCatalog(value), nil
}

func (catalog *Catalog) save(value StudyCatalog) error {
	if value.RecentStudies == nil {
		value.RecentStudies = []RecentStudyEntry{}
	}

	if err := os.MkdirAll(catalog.rootDir, 0o755); err != nil {
		return contracts.Internal(
			fmt.Sprintf("failed to create catalog directory %s: %v", catalog.rootDir, err),
		)
	}

	payload, err := json.MarshalIndent(value, "", "  ")
	if err != nil {
		return contracts.Internal(fmt.Sprintf("serialize study catalog: %v", err))
	}

	if err := os.WriteFile(catalog.path, payload, 0o644); err != nil {
		return contracts.Internal(
			fmt.Sprintf("failed to write study catalog %s: %v", catalog.path, err),
		)
	}

	catalog.cache = cloneStudyCatalog(value)
	catalog.loaded = true

	return nil
}

func (catalog *Catalog) corruptPath() string {
	extension := filepath.Ext(catalog.path)
	if extension == "" {
		return catalog.path + ".corrupt"
	}

	return strings.TrimSuffix(catalog.path, extension) + ".corrupt" + extension
}

func persistedMeasurementScale(
	scale *contracts.MeasurementScale,
) *PersistedMeasurementScale {
	if scale == nil {
		return nil
	}

	return &PersistedMeasurementScale{
		RowSpacingMM:    scale.RowSpacingMM,
		ColumnSpacingMM: scale.ColumnSpacingMM,
		Source:          scale.Source,
	}
}

func emptyStudyCatalog() StudyCatalog {
	return StudyCatalog{
		RecentStudies: []RecentStudyEntry{},
	}
}

func cloneStudyCatalog(value StudyCatalog) StudyCatalog {
	recentStudies := make([]RecentStudyEntry, len(value.RecentStudies))
	for index, entry := range value.RecentStudies {
		recentStudies[index] = entry
		if entry.MeasurementScale != nil {
			scale := *entry.MeasurementScale
			recentStudies[index].MeasurementScale = &scale
		}
	}

	return StudyCatalog{
		RecentStudies: recentStudies,
	}
}
